src/driver/kvm.rs

```diff
@@ -474,6 +455,1009 @@ mod tests {
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+        kvm.expect_pause_ev = 5;
+        let result = Kvm::resume(&mut kvm);
+
+        assert!(result.is_err(), "Expected error, got ok instead!");
+    }
+
+    #[test]
+    fn test_resume_vcpu_fails_if_kvmi_reply_fails() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        kvmi_mock
+            .expect_wait_and_pop_event()
+            .times(1)
+            .returning(|_| {
+                Ok(Some(KVMiEvent {
+                    vcpu: 0,
+                    ev_type: KVMiEventType::PauseVCPU,
+                    ffi_event: std::ptr::null_mut(),
+                }))
+            });
+        kvmi_mock.expect_reply().times(1).returning(|_, _| {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "something went wrong",
+            ))
+        });
+
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+        kvm.expect_pause_ev = 5;
+        let result = Kvm::resume(&mut kvm);
+
+        assert!(result.is_err(), "Expected error, got ok instead!");
+    }
+
+    #[test]
+    fn test_read_physical_succeeds() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        kvmi_mock
+            .expect_read_physical()
+            .times(1)
+            .returning(|_, _| Ok(()));
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+        let paddr: u64 = 0;
+        let mut buffer: [u8; 4096] = [0; 4096];
+        let result = Kvm::read_physical(&mut kvm, paddr, &mut buffer);
+        assert!(result.is_ok(), "Expected ok, got error instead!");
+    }
+
+    #[test]
+    fn test_read_physical_fails_if_kvmi_read_physical_fails() {
+        let mut kvmi_mock = MockKVMi::default();
+        setup_kvmi_mock(&mut kvmi_mock);
+        kvmi_mock.expect_read_physical().times(1).returning(|_, _| {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "something went wrong",
+            ))
+        });
+        let mut kvm = Kvm::new(
+            "some_vm",
+            kvmi_mock,
+            Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())),
+        )
+        .expect("Failed to create driver");
+        let paddr: u64 = 0;
+        let mut buffer: [u8; 4096] = [0; 4096];
+        let result = Kvm::read_physical(&mut kvm, paddr, &mut buffer);
+        assert!(result.is_err(), "Expected error, got ok instead!");
+    }
+
+    #[test]
+    fn test_write_physical_succeeds() {
```

Comment:
Same principle as in read_physical, just the other way around. Expect some specific content in the mocked call.